logexp/cmd/main.go

```diff
@@ -0,0 +1,253 @@
+			glog.Warningf("Failed to record journalctl logs: %v", err)
+		}
+	}
+}
+
+// Copy logfiles specific to this node based on its type, cloud-provider, system services, etc
+// to a temporary directory. Also create logfiles for systemd services if journalctl is present.
+// We do not expect this function to see an error.
+func prepareLogfiles(logDir string) {
+	glog.Infof("Preparing logfiles relevant to this node")
+	logfiles := make([]string, 0)
+	var isMaster bool
+
+	// Select node-type (master/node) specific logs.
+	if _, err := os.Stat(localLogPath + "/kube-apiserver.log"); err == nil {
+		glog.Infof("VM identified as kubernetes master")
+		isMaster = true
+		logfiles = append(logfiles, masterLogs...)
+	} else {
+		glog.Infof("VM identified as kubernetes node")
+		isMaster = false
+		logfiles = append(logfiles, nodeLogs...)
+	}
+
+	// Select cloud-provider specific logs.
+	switch kubernetesProvider {
+	case "gce", "gke":
+		logfiles = append(logfiles, gceLogs...)
+	case "kubemark":
+		logfiles = append(logfiles, gceLogs...)  // TODO(shyamjvs): Make this variable wrt kubemark's real provider.
+		if ! isMaster && enableHollowNodeLogs {
+			logfiles = append(logfiles, kubemarkLogs...)
+		}
+	case "aws":
+		logfiles = append(logfiles, awsLogs...)
+	default:
+		glog.Warningf("Unknown cloud provider '%v' provided, skipping any provider specific logs", kubernetesProvider)
+	}
+
+	// Select system/service specific logs.
+	if _, err := os.Stat("/workspace/etc/systemd/journald.conf"); err == nil {
+		glog.Infof("Journalctl found on host. Collecting systemd logs")
+		createSystemdLogfiles(logDir, isMaster)
+	} else {
+		glog.Infof("Journalctl not found on host (%v). Collecting supervisord logs instead", err)
+		logfiles = append(logfiles, kernelLog)
+		logfiles = append(logfiles, initdLogs...)
+		logfiles = append(logfiles, supervisordLogs...)
+	}
+
+	// Copy all the logfiles that exist, to logDir.
+	for _, logfile := range logfiles {
+		logfileFullPath := filepath.Join(localLogPath, logfile + ".log*")  // Append .log* to copy rotated logs too.
+		cmd := exec.Command("/bin/sh", "-c", fmt.Sprintf("cp %v %v", logfileFullPath, logDir))
+		if err := cmd.Run(); err != nil {
+			glog.Warningf("Failed to copy any logfiles with pattern '%v': %v", logfileFullPath, err)
+		}
+	}
+}
+
+// Effeciently upload all logfiles inside logDir to GCS.
+func uploadLogfilesToGCS(logDir string) error {
+	cmd := exec.Command("/bin/sh", "-c", fmt.Sprintf("ls %v/*", logDir))
+	if output, err := cmd.Output(); err != nil {
+		glog.Errorf("%v", )
+		return fmt.Errorf("Could not list any logfiles: %v", err)
+	} else {
+		glog.Infof("List of logfiles available: %v", string(output))
+	}
+
+	// Upload the files with compression and parallelism (speeds up on good network)
+  	// along with setting their ACL to make them publicly readable.
+  	gcsLogPath := filepath.Join(gcsArtifactsPath, nodeName)
+  	if ! strings.HasPrefix(gcsLogPath, "gs://") {
```

Comment:
Good news, got rid of the hack! It turns out that this issue of multiple slash simplification only occurs when we try calling the binary directly (not as part of container command), as then bash is in play. However in case of container too the same issue occurred but it was because of a different reason (which is filepath.Join() also does that simplification). Changed it now.